### PR TITLE
make enums containing ByteBuffer fit in three words

### PR DESCRIPTION
**Pull request by weissi**
CC @tanner0101 & @helje5 

only really makes sense together with #351

Motivation:

NIO's dynamic pipeline comes with one performance caveat: If your type
isn't specialised in NIOAny (only types in the NIO module can be) and
it's over 3 words (24 bytes on 64-bit platforms), you'll suffer an
allocation for every time you box it in NIOAny. Very unfortunately, both
ByteBuffer and FileRegion were exactly 3 words wide which means that any
enum containing those would be wider than 3 words. Worse, both
HTTP{Request,Response}Head contained types that were wider than 3 words.
The best solution to this problem is to shrink ByteBuffer and FileRegion
to just under 3 words and that's exactly what this PR is doing. That is
slightly tricky as ByteBuffer was already bit packed fairly well
(reader/writer indices & slice begin/end were stored as a UInt32). The
trick we employ now is to store the slice beginning in a UInt24 and the
file region reader index in a UInt56. That saves one byte for both
ByteBuffer and FileRegion with very moderate tradeoffs: The reader index
in a file now needs to be within 64 PiB (peta bytes) and a byte buffer
slice beginning must start within 16 MiB (mega bytes). Note: The
reader/writer indices as well as slice ends are _not_ affected and can
still be within 4 GiB. Clearly no one would care about the restrictions
for FileRegions in the real world but we might hit the ByteBuffer slice
beginning limit in which case the slice would be copied out. But
given that slices are mostly used to slice off headers in network
protocols, 16 MiB should be _plenty_.

Norman was kind enough to measure the perf differences:

master (before this PR):
```
$ wrk -c 256 -d 10s -t 4 -s ~/Downloads/pipeline-many.lua -H "X-Host: SomeValue" -H "Host: swiftnio.io" -H "ThereAreEvenMoreHeaders: AndMoreValues" http://localhost:8888
Running 10s test @ http://localhost:8888
  4 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    93.02ms  158.13ms   1.97s    93.07%
    Req/Sec   112.79k    26.05k  258.75k    84.50%
  4501098 requests in 10.04s, 214.63MB read
  Socket errors: connect 0, read 111, write 0, timeout 89
Requests/sec: 448485.61
Transfer/sec:     21.39MB
```

after this PR:
```
$ wrk -c 256 -d 10s -t 4 -s ~/Downloads/pipeline-many.lua -H "X-Host: SomeValue" -H "Host: swiftnio.io" -H "ThereAreEvenMoreHeaders: AndMoreValues" http://localhost:8888
Running 10s test @ http://localhost:8888
  4 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   107.53ms  206.56ms   1.99s    90.55%
    Req/Sec   124.15k    26.56k  290.41k    89.25%
  4952904 requests in 10.03s, 236.17MB read
  Socket errors: connect 0, read 161, write 0, timeout 22
Requests/sec: 493852.65
Transfer/sec:     23.55MB
```

so we see a nice 10% improvement and allocations go down by 2, too:

```
23:21:44 info: 1000_reqs_1_conn: allocations not freed: 0
23:21:44 info: 1000_reqs_1_conn: total number of mallocs: 67832
```

from

```
19:08:57 info: 1000_reqs_1_conn: allocations not freed: 0
19:08:57 info: 1000_reqs_1_conn: total number of mallocs: 69834
```

Modifications:

- shrank ByteBuffer by 1 byte, making it 23 bytes in total
- shrank FileRegion by 1 byte, making it 23 bytes in total
- CoW boxed parts of HTTP{Request,Response}Head to make them fit within 2 words
- added @_inlineable to NIOAny where appropriate to not suffer boxed existentials in different places
- added tests for the new edge cases

Result:

- more speed
- fewer allocations

### CODE REVIEW

File: Sources/NIO/ByteBuffer-core.swift [hunk 1]
```diff
@@ -29,6 +29,86 @@ let sysFree: @convention(c) (UnsafeMutableRawPointer?) -> Void = free
     }
 #endif
 
+struct SadUInt24: ExpressibleByIntegerLiteral {
```
Comment (normanmaurer) [inline Sources/NIO/ByteBuffer-core.swift]: Haha... `Sad...` ;)

File: Sources/NIO/ByteBuffer-core.swift [hunk 2]
```diff
@@ -29,6 +29,86 @@ let sysFree: @convention(c) (UnsafeMutableRawPointer?) -> Void = free
...
+    @_versioned var b3: UInt8
+
+    init(integerLiteral value: UInt16) {
+        self.b1 = UInt8(truncatingIfNeeded: (value & 0xff      ) >> 0 )
+        self.b2 = UInt8(truncatingIfNeeded: (value & 0xff_00   ) >> 8 )
+        self.b3 = 0
+    }
+
+    init?(_ value: UInt32) {
+        guard value & 0xff_00_00_00 == 0 else {
```
Comment (normanmaurer) [inline Sources/NIO/ByteBuffer-core.swift]: TIL that you can use `_` to make it more readable

File: Sources/NIOHTTP1/HTTPTypes.swift [hunk 3]
```diff
@@ -19,11 +19,52 @@ let headerSeparator: StaticString = ": "
 
 /// A representation of the request line and header fields of a HTTP request.
 public struct HTTPRequestHead: Equatable {
+    private class _Storage {
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPTypes.swift]: I guess it not really makes a difference but use `final` (for all internal classes) ?

File: Sources/NIOHTTP1/HTTPTypes.swift [hunk 4]
```diff
@@ -19,11 +19,52 @@ let headerSeparator: StaticString = ": "
...
+        var version: HTTPVersion
+
+        init(method: HTTPMethod, rawURI: URI, version: HTTPVersion) {
+            self.method = method
+            self.rawURI = rawURI
+            self.version = version
+        }
+
+        func copy() -> _Storage {
+            return _Storage(method: self.method, rawURI: self.rawURI, version: self.version)
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPTypes.swift]: Can you either always use `.init` or always `_Storage` just to make it consistent ?

File: Sources/NIO/ByteBuffer-core.swift [hunk 5]
```diff
@@ -486,8 +584,19 @@ public struct ByteBuffer {
         }
         let index = _toIndex(index)
         let length = _toCapacity(length)
+        let sliceBeginIndex = self._slice.lowerBound + index
+
+        guard sliceBeginIndex <= ByteBuffer.Slice.maxSupportedLowerBound else {
+            // the slice's begin is past the maximum supported slice begin value (16 MiB) so the only option we have
+            // is copy the slice into a fresh buffer and the slice begin will then be at index 0.
```
Comment (normanmaurer) [inline Sources/NIO/ByteBuffer-core.swift]: nit: s/and the/. The/

File: Sources/NIO/ByteBuffer-core.swift [hunk 6]
```diff
@@ -29,6 +29,97 @@ let sysFree: @convention(c) (UnsafeMutableRawPointer?) -> Void = free
...
+        newValue |= Int(value.b3) << 16
+        self = newValue
+    }
+}
+
+extension _UInt24 {
+    init(_ value: UInt32) {
+        precondition(value & 0xff_00_00_00 == 0, "value \(value) too large for _UInt24")
+        self.b12 = UInt16(truncatingIfNeeded: (value &    0xff_ff) >> 0 )
+        self.b3  =  UInt8(truncatingIfNeeded: (value & 0xff_ff_00) >> 16)
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-core.swift]: You can save yourself a masking operation here. You already precondition that the top byte is zero, so with that knowledge in place you can just shift right by 16 bits and store into `UInt8`.

File: Sources/NIO/ByteBuffer-core.swift [hunk 7]
```diff
@@ -29,6 +29,97 @@ let sysFree: @convention(c) (UnsafeMutableRawPointer?) -> Void = free
     }
 #endif
 
+@_versioned
+struct _UInt24: ExpressibleByIntegerLiteral {
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-core.swift]: Can we get a doc comment here please?

File: Sources/NIO/ByteBuffer-core.swift [hunk 8]
```diff
@@ -29,6 +29,97 @@ let sysFree: @convention(c) (UnsafeMutableRawPointer?) -> Void = free
...
+}
+
+extension _ByteBufferSlice: Equatable {
+    static func ==(_ lhs: _ByteBufferSlice, _ rhs: _ByteBufferSlice) -> Bool {
+        return lhs._begin == rhs._begin && lhs.upperBound == rhs.upperBound
+    }
+}
+
+@_versioned
+struct _ByteBufferSlice {
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-core.swift]: Similarly, can we get a doc comment here please?

File: Sources/NIO/ByteBuffer-core.swift (hunk 6 above)
Comment (Lukasa) [reply Sources/NIO/ByteBuffer-core.swift]: Also as noted in the chat, the mask isn't really right anyway.

File: Sources/NIO/ByteBuffer-core.swift [hunk 9]
```diff
@@ -486,8 +577,19 @@ public struct ByteBuffer {
         }
         let index = _toIndex(index)
         let length = _toCapacity(length)
+        let sliceBeginIndex = self._slice.lowerBound + index
+
+        guard sliceBeginIndex <= ByteBuffer.Slice.maxSupportedLowerBound else {
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-core.swift]: Nit: the word "begin" here is odd. Start might be better.

File: Sources/NIO/FileRegion.swift [hunk 10]
```diff
@@ -12,6 +12,68 @@
 //
 //===----------------------------------------------------------------------===//
 
+struct _UInt56: ExpressibleByIntegerLiteral {
```
Comment (Lukasa) [inline Sources/NIO/FileRegion.swift]: Might be nice to consolidate our helper integers into a single source file.

File: Sources/NIO/FileRegion.swift [hunk 11]
```diff
@@ -12,6 +12,68 @@
...
+
+    static let min: _UInt56 = 0
+}
+
+extension _UInt56 {
+    init(_ value: UInt64) {
+        precondition(value & 0xff_00_00_00_00_00_00_00 == 0)
+        self.init(b1234: UInt32(truncatingIfNeeded: (value &          0xff_ff_ff_ff) >> 0 ),
+                  b56:   UInt16(truncatingIfNeeded: (value &    0xff_ff_00_00_00_00) >> 32),
+                  b7:     UInt8(truncatingIfNeeded: (value & 0xff_00_00_00_00_00_00) >> 48))
```
Comment (Lukasa) [inline Sources/NIO/FileRegion.swift]: Given your earlier precondition, this mask is unnecessary.

File: Sources/NIO/ByteBuffer-core.swift (hunk 6 above)
Comment (weissi) [reply Sources/NIO/ByteBuffer-core.swift]: oops! thanks

File: Sources/NIO/FileRegion.swift (hunk 10 above)
Comment (weissi) [reply Sources/NIO/FileRegion.swift]: fair enough, will do that

File: Sources/NIO/IntegerTypes.swift [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        var newValue: Int = 0
+        newValue  = Int(value.b12)
+        newValue |= Int(value.b3) << 16
+        self = newValue
+    }
+}
+
+extension _UInt24 {
+    init(_ value: UInt32) {
+        precondition(value & 0xff_00_00_00 == 0, "value \(value) too large for _UInt24")
```
Comment (normanmaurer) [inline Sources/NIO/IntegerTypes.swift]: @weissi should this be an assert as there is no way for the user to actual construct this ? I think we usually use `assert` for things that would only happen because of a bug in NIO itself

File: Sources/NIO/IntegerTypes.swift [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    static var max: _UInt56 {
+        return .init(b1234: .max, b56: .max, b7: .max)
+    }
+    
+    static let min: _UInt56 = 0
+}
+
+extension _UInt56 {
+    init(_ value: UInt64) {
+        precondition(value & 0xff_00_00_00_00_00_00_00 == 0)
```
Comment (normanmaurer) [inline Sources/NIO/IntegerTypes.swift]: @weissi should this be an assert as there is no way for the user to actual construct this ? I think we usually use `assert` for things that would only happen because of a bug in NIO itself

Also consider adding a message as well

File: Tests/NIOTests/ByteBufferTest.swift [hunk 14]
```diff
@@ -1214,6 +1214,87 @@ class ByteBufferTest: XCTestCase {
...
+        var buf = self.allocator.buffer(capacity: totalBufferSize)
+        for _ in 0..<5 {
+            var fresh = fourMBBuf
+            buf.write(buffer: &fresh)
+        }
+        XCTAssertEqual(totalBufferSize, buf.readableBytes)
+
+        let offset = Int(_UInt24.max) + 1
+        let expectedReadablebytes = totalBufferSize - offset
+        let slice = buf.getSlice(at: offset, length: expectedReadablebytes)!
```
Comment (Lukasa) [inline Tests/NIOTests/ByteBufferTest.swift]: Minor nit: it might be better if we sliced the same size as the test above, to isolate the test differences to only being the slice start index, rather than the slice sizes.

Comment (weissi) [reply Tests/NIOTests/ByteBufferTest.swift]: addressed and improved test (by marking some 'important' positions with different byte values)

**Comment by normanmaurer** [comment]
@weissi please merge once CI completes
